src/snabbdom.ts

```diff
@@ -60,125 +95,111 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
         api.removeChild(parent, childElm);
       }
     };
   }
 
   function createElm(vnode: VNode, insertedVnodeQueue: VNodeQueue): Node {
-    let i: any, data = vnode.data;
-    if (data !== undefined) {
+    let i: any;
+    const {data} = vnode;
+    if (!isUndef(data)) {
       if (isDef(i = data.hook) && isDef(i = i.init)) {
         i(vnode);
-        data = vnode.data;
       }
     }
-    let children = vnode.children, sel = vnode.sel;
-    if (sel === '!') {
-      if (isUndef(vnode.text)) {
-        vnode.text = '';
-      }
-      vnode.elm = api.createComment(vnode.text as string);
-    } else if (sel !== undefined) {
-      // Parse selector
-      const hashIdx = sel.indexOf('#');
-      const dotIdx = sel.indexOf('.', hashIdx);
-      const hash = hashIdx > 0 ? hashIdx : sel.length;
-      const dot = dotIdx > 0 ? dotIdx : sel.length;
-      const tag = hashIdx !== -1 || dotIdx !== -1 ? sel.slice(0, Math.min(hash, dot)) : sel;
-      const elm = vnode.elm = isDef(data) && isDef(i = (data as VNodeData).ns) ? api.createElementNS(i, tag)
-                                                                               : api.createElement(tag);
-      if (hash < dot) elm.setAttribute('id', sel.slice(hash + 1, dot));
-      if (dotIdx > 0) elm.setAttribute('class', sel.slice(dot + 1).replace(/\./g, ' '));
+    if (isElementVNode(vnode)) {
+      const {data, tag} = vnode;
+      const elm = vnode.elm = isDef(i = data.ns) ? api.createElementNS(i, tag)
+                                                 : api.createElement(tag);
+      if (isDef(i = data.hook) && isDef(i.create)) i.create(emptyNode, vnode);
       for (i = 0; i < cbs.create.length; ++i) cbs.create[i](emptyNode, vnode);
-      if (is.array(children)) {
+      const { children } = vnode;
+      if (isArray(children)) {
         for (i = 0; i < children.length; ++i) {
           const ch = children[i];
-          if (ch != null) {
-            api.appendChild(elm, createElm(ch as VNode, insertedVnodeQueue));
+          if (isVNode(ch)) {
+            api.appendChild(elm, createElm(ch, insertedVnodeQueue));
           }
         }
-      } else if (is.primitive(vnode.text)) {
+      } else if (!isUndef(vnode.text)) {
         api.appendChild(elm, api.createTextNode(vnode.text));
       }
-      i = (vnode.data as VNodeData).hook; // Reuse variable
-      if (isDef(i)) {
-        if (i.create) i.create(emptyNode, vnode);
-        if (i.insert) insertedVnodeQueue.push(vnode);
-      }
+      if (isDef(i = data.hook) && isDef(i.insert)) insertedVnodeQueue.push(vnode);
+    } else if (isTextVNode(vnode)) {
+      vnode.elm = api.createTextNode(vnode.text);
+    } else if (isCommentVNode(vnode)) {
+      vnode.elm = api.createComment(vnode.text);
+    } else if (isFragmentVNode(vnode)) {
+      vnode.elm = api.createFragment();
     } else {
-      vnode.elm = api.createTextNode(vnode.text as string);
+      throw new TypeError();
```

Comment:
as I mentioned above, no more fallback mechanism, it is a error if you provide a weird constructed vnode.